Trigger switch implementation with laser diode, lcd backlight settings

krishnadurai opened the pull request:
This PR implements the Trigger switch with laser diode, lcd backlight settings.

Resolves #8: Resets moving average on trigger release.

- On trigger being ON: Temperature is recorded with IR sensor. LCD and Laser Diode are enabled.
- On trigger being OFF: Temperature is recording is disabled. LCD and Laser Diode are disabled.
- Hardware pins match specifications published
- Macros are being used instead of constants for pins and code constants.

krishnadurai committed:
Moving averages computed for stable temperature reading.

TODOs:
1. Compute final reading with this.
2. Change type of values to float for averages.

krishnadurai committed:
Removes heap usage
Clears up bugs in code

krishnadurai committed:
Trigger switch implemented for loop()

kamatajinkya reviewed (dismissed):
Looks good for a first cut. Need to look into sleep modes

yashtambi commented on TempSensor.ino:
```diff
@@ -142,27 +144,121 @@ uint16_t MLX90614::read16(uint8_t a) {
-  delay(500);
+  int temp_celcius;
+  int temp_fahrenheit;
+  switch (triggerState){
```
Is this code for simulation? 
How are you reading the trigger switch?

yashtambi commented on TempSensor.ino:
```diff
@@ -142,27 +144,121 @@ uint16_t MLX90614::read16(uint8_t a) {
+  int temp_fahrenheit;
+  switch (triggerState){
+    case true:
+      // Get temperature readings
```
Also need to 'enable laser diode' in 'true' case

krishnadurai replied:
Done

yashtambi commented on TempSensor.ino:
```diff
@@ -142,27 +144,121 @@ uint16_t MLX90614::read16(uint8_t a) {
+      lcd.display();
+      break;
+    case false:
+      lcd.noDisplay();
```
Also need to disable Laser Diode in 'false' case

krishnadurai replied:
Done

krishnadurai committed:
Defines pins for Custom LCD backlight and brightness
Defines pins for Trigger Switch
Defines pins for Laser Diode
Coverts all pin constants and code constants to Macros

krishnadurai commented:
@yashtambi @kamatajinkya I have updated the PR title and [description](https://github.com/CaringIndians/thermalsensor/pull/6#issue-394992229).

This is ready for another round of reviews.

krishnadurai commented on TempSensor.ino:
```diff
@@ -144,15 +144,97 @@ uint16_t MLX90614::read16(uint8_t a) {
+  public:
+    void begin(){
+      pinMode(TRIGGER_PIN, INPUT);
+      digitalWrite(TRIGGER_PIN, HIGH);
```
We are not able to set this to 'LOW' as the simulated circuit doesn't start up.

kamatajinkya replied:
Not sure what you mean by simulated circuit does not start up. The code looks fine to me.

krishnadurai replied:
Got it now

krishnadurai commented on TempSensor.ino:
```diff
@@ -210,51 +292,73 @@ class MovingAverage {
+  int temp_fahrenheit;
+
+  switch (trigger.readTriggerStatus()){
+    case LOW:
```
Since we are not able to start the trigger with a setting of LOW in
https://github.com/CaringIndians/thermalsensor/pull/6/files#diff-a7e9cd7b80b8f4bf39578dbe1daea2b1R198

We need to treat LOW as the case when the temperature will be measured.

Is my understanding wrong here?

@yashtambi @kamatajinkya

kamatajinkya replied:
No, you are right. In the setup code, the line `digitalWrite(TRIGGER_PIN, DIGITAL_HIGH)` will setup and internal pull up making sure that the input is always high, except when you press the trigger button.

krishnadurai replied:
Thanks for clarifying

kamatajinkya reviewed (approved):
Looks good to me

yashtambi commented on TempSensor.ino:
```diff
@@ -210,51 +292,73 @@ class MovingAverage {
+      temp_fahrenheit = mlx.readObjectTempF();
+      // Populate temperature values for moving average
+      if (FLAG_MOVING_AVG_ENABLE == true) {
+        avg_celsius.pushValue(temp_celcius);
```
Not necessary for now but you can probably have just 1 'if(FLAG_MOVING_AVG_ENABLE)' and have the code for displaying data in C and F in the same place. 
In other words, having a separate check for MOVING_AVG_ENABLE for C and F is not required
